File: benchmarks/remediation/fixtures/python-path-flask-open/tests/verify.py
"""Two invocation modes.

  python3 tests/verify.py --exploit | --behavior   Sandbox mode: checks the app.py in this tree.
  python3 tests/verify.py <original> <repaired>    Harness mode: compares the checked-in trees.
"""
import builtins
import importlib.util
import io
import json
import os
import sys
import types

ATTACK = "../../etc/passwd"
LEGITIMATE = "2024/march.pdf"
BASE = "/srv/invoices"

class HTTPError(Exception):
    def __init__(self, code):
        Exception.__init__(self, str(code))
        self.code = code


class FakeRequest:
    def __init__(self):
        self.args = {}


REQUEST = FakeRequest()
ROUTES = {}
READS = []


def route(rule, **kwargs):
    def decorate(view):
        ROUTES[rule] = view
        return view

    return decorate


class FakeFlask:
    def __init__(self, name):
        self.name = name

    route = staticmethod(route)


def send_file(target, **kwargs):
    with open_recorded(target) as handle:
        return handle.read()


def open_recorded(target, mode="rb"):
    READS.append(str(target))
    return io.BytesIO(b"%PDF-1.4 fixture")


flask = types.ModuleType("flask")
flask.Flask = FakeFlask
flask.request = REQUEST
flask.abort = lambda code, *a, **k: (_ for _ in ()).throw(HTTPError(code))
flask.send_file = send_file
sys.modules.setdefault("flask", flask)
builtins.open = lambda target, mode="r", *a, **k: open_recorded(target, mode)

def load(path, name="app_under_test"):
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def local_path():
    return os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "app.py")


def fetch(module, name):
    del READS[:]
    REQUEST.args = {"name": name}
    try:
        return module and ROUTES["/invoices"](), None
    except HTTPError as error:
        return None, error.code


def exploit():
    module = load(local_path())
    fetch(module, ATTACK)
    escaped = [item for item in READS if not os.path.normpath(item).startswith(BASE + os.sep)]
    if escaped:
        sys.stderr.write("vulnerability present: the resolved path escapes the base directory\n")
        sys.exit(1)
    sys.exit(0)


def behavior():
    module = load(local_path())
    body, code = fetch(module, LEGITIMATE)
    assert code is None and body is not None
    assert READS and os.path.normpath(READS[0]) == os.path.join(BASE, LEGITIMATE)
    sys.exit(0)


def compare(original_path, repaired_path):
    original = load(original_path, "original")
    ROUTES.clear()
    load(original_path, "original_again")
    fetch(original, ATTACK)
    assert os.path.normpath(READS[0]) == "/etc/passwd"
    ROUTES.clear()
    repaired = load(repaired_path, "repaired")
    body, code = fetch(repaired, ATTACK)
    assert READS == [] and code == 400
    body, code = fetch(repaired, LEGITIMATE)
    assert code is None and os.path.normpath(READS[0]) == os.path.join(BASE, LEGITIMATE)
    sys.stdout.write(json.dumps({"vulnerability_observed": True, "behavior_preserved": True}))


if __name__ == "__main__":
    mode = sys.argv[1] if len(sys.argv) > 1 else ""
    if mode == "--exploit":
        exploit()
    elif mode == "--behavior":
        behavior()
    else:
        compare(sys.argv[1], sys.argv[2])

File: benchmarks/remediation/fixtures/js-path-destructured-parameter/tests/verify.js
// Two invocation modes.
//   node tests/verify.js --exploit | --behavior   Sandbox mode: checks the app.js in this tree.
//   node tests/verify.js <original> <repaired>    Harness mode: compares the checked-in trees.
//
// fs.readFile is replaced with a recorder, so no file is opened: the exploit check passes only
// when a traversal name records no read at all.
const assert = require('node:assert/strict');
const path = require('node:path');
const fs = require('node:fs');

const ATTACK = '../../etc/passwd';
const LEGITIMATE = 'handbook/intro.md';
const BASE = '/srv/docs';
const READS = [];

fs.readFile = (target, encoding, callback) => {
  READS.push(String(target));
  const done = typeof encoding === 'function' ? encoding : callback;
  if (typeof done === 'function') done(null, '# fixture document');
};

function load(target) {
  READS.length = 0;
  return require(target);
}

function loadLocal() {
  return load(path.resolve(__dirname, '..', 'app.js'));
}

// A refusal reaches the caller either way a repair can signal one: through the callback, or by
// throwing. The fixture cares that the name was refused and nothing was read, not which. Without
// the catch a throwing repair would crash this script and be scored as the vulnerability.
function fetch(module, name) {
  READS.length = 0;
  let error = null;
  let body = null;
  try {
    module.readDocument({ name }, (failure, content) => { error = failure; body = content; });
  } catch (thrown) {
    error = thrown;
  }
  return { error, body };
}

function exploit() {
  fetch(loadLocal(), ATTACK);
  const escaped = READS.filter((item) => !path.resolve(item).startsWith(`${BASE}${path.sep}`));
  if (escaped.length > 0) {
    process.stderr.write('vulnerability present: the resolved path escapes the served directory\n');
    process.exit(1);
  }
  process.exit(0);
}

function behavior() {
  const result = fetch(loadLocal(), LEGITIMATE);
  assert.equal(result.error, null);
  assert.equal(result.body, '# fixture document');
  assert.deepEqual(READS, [path.join(BASE, LEGITIMATE)]);
  process.exit(0);
}

function compare(originalPath, repairedPath) {
  fetch(load(originalPath), ATTACK);
  assert.deepEqual(READS, ['/etc/passwd']);
  const repaired = load(repairedPath);
  const blocked = fetch(repaired, ATTACK);
  assert.deepEqual(READS, []);
  assert.ok(blocked.error instanceof Error);
  const allowed = fetch(repaired, LEGITIMATE);
  assert.equal(allowed.error, null);
  assert.deepEqual(READS, [path.join(BASE, LEGITIMATE)]);
  process.stdout.write(JSON.stringify({ vulnerability_observed: true, behavior_preserved: true }));
}

const mode = process.argv[2];
if (mode === '--exploit') exploit();
else if (mode === '--behavior') behavior();
else compare(process.argv[2], process.argv[3]);
